### DISCUSSION TITLE
How to get Sveltekit to refresh the data when going from one [slug] to another?

### DISCUSSION THREAD

**Discussion by hubert-ix** [sveltejs/kit#9533] [Q&A] [answered]
It's probably a common question but I couldn't find the answer.

I have a `/routes/[slug]` folder, and a pretty simple case:

`+page.js`

``` javascript
export async function load({ fetch, params, url }) {
  let item = await fetchItemFromDb();
  return {item }
}
```

`+page.svelte`

``` svelte
<script>
export let data;

let item = data.item;
</script>

{data.item.title}
```

Then I have a navigation bar with links to `/test1` and `/test2`.

I go to `/test1`, and I get the data for that URL. Then I click on `/test2`, but sveltekit doesn't refresh anything. 

`+page.js` is actually called again and the new data is passed on to `+page.svelte` (I don't even have to use `invalidate()`), but the new data is not rendered on the page.

So if I update `+page.svelte` to this:

``` svelte
<script>
export let data;

let item = data.item;

$: if (data) {
  console.log(data.item.title)
}
</script>

{item.title}
```

When I click on `/test2`, I get the new title in the console, but not rendered on the page.

How can I get it to render the new data?

**Comment by hubert-ix** [comment, marked as answer]
Ah, duh... I simply have to declare `item` with a reactive statement, otherwise sveltekit doesn't know it has changed...

`+page.svelte`

``` svelte
<script>
export let data;

$: item = data.item;
</script>

{item.title}
```

**Reply by eallenOP** [reply]
This helped me with the same problem. So simple! Thank you :)

**Reply by dangelomedinag** [reply]
when you use the `params` argument to the load function, Sveltekit automatically 'detects' that the parameter has changed, and updates the value of data (most commonly the `params.slug` is used to get dynamic data).  In your case, since you don't use `params.slug` inside the load function, you have to make a reactive statement for the changes to be rendered.

**Reply by hubert-ix** [reply]
Oh, very interesting, I didn't know that. Thanks for the tip!

**Reply by eallenOP** [reply]
I'm using the params value in my load function but I still got the OP issue. The reactive statement in +page.svelte fixed it for me.

src/routes/location/[location]/+page.server.js
```js
export async function load( {params} ) {
  
  const url = 'https://api.spacetraders.io/v2/systems/X1-DF55/waypoints/' + params.location;

  const response = await fetch(url, options); 
  const location_details = await response.json(); 

  if (location_details) {
    return {
      location_details,
    };
  }
  return {
    status: 404,
  };
}
```

**Reply by sandrovilar** [reply]
I'm was using _goto_ to the same route (with different params) and getting the length of multiple arrays (in data) was always the same (previous request) to perform some layout adjustments, using a reactive statement solved the problem, thanks!

**Comment by dangelomedinag** [comment]
@eallenOP  I did a [minimal repro](https://www.sveltelab.dev/#code=N4IgdADgTgpgLnAljKUDGIBcoBmiA2MWoaA9gHZwyUDOxIA1ouQCZYg0BuM%2BV%2BAhgCMAtAFdEADn4gANCEGicOFFgDaARgBMAZhnqADDICcM7QBY96gOyWArHv3qZEi0ZtGJd0xduftmywA2Swt1awcnM1DDE3MvA3swwwNdA08DJwknMJsw7PVgjO8ZX1MAsML1UPCiqIdjYoqQ41z9RP10tstk-V1A3LCHEwMTN2LS-0aDS3yepzi66djq8qqHROmmiVb25IKHCziJgLMTM09zmUWYqfKOy1bUrqSvI78Ao3mV5qKmv%2BfHM0zBs7ukar11t0rhsNp1svd-l9jLprg04mFQQ8HE8NoYJE8nGMwq5hvoLIFOhtqvY3mVjEizME4kYUdE0dV8nCfmsEg45q8fO96XodHT0fo7u0uWEaYKxRYzCdKrCIg4wYVAQtDgF0YMMUFIXkoTLxkKtLpJrqPgNqariab5fF1NKEfsinFJloaeVNPYWfUvQ4ALoAXzDcgg-DQDH4AHMYGAAFY0Cj0PCEehkSjUOB0bCMZhsTAcbi8HhCMSSaRyBRKFSYDSipZTGKtUlI463bnwjWE4b7LX1ZaWaU89TtDazB2WhUXYKXRlXYeo4f6%2B28sKeIluzVyz1usetSJs1duzT5O4doXi-vwvatLmD5u6v396Igg0m2meps3YceLENwRfRDG-HVqkpT9NQ%2BMEdjVPlihXdkEN%2BZ1p3AkdDR3eY7n5TdEJPZC1x5QF8KA-Jh07XVIMROliOiTlDVI9RVxuQMkPFPC0IpO0tF4nouJiUl2nyWUSjNX8GmHQJbTI0jXV7BxlSudxWlbSxVkU%2B1gRaQCDyvR0jB8QxFwXec-AVQjpNktDfgUu19GUsxWjgpowOKIylyuYJbBRcyCKHIi%2BPXZjKg1ETmnc5ljO8kp538p8-y7EL8jCpj8gVD80kAiEN2mHSTTk%2BF1VVKLXBixdbHirzEqkqYVTskr3T3DDPNsSIfN0WwlSuCyrisqYaOw9CPJ8HrTli7rGjY30A1mgxQ3DDh0HoFhEFgNA4FIKAAE96H4CAIDAAALOAAFt8DTAgiHzLMqFoegmFYdguB4PgK3EKRZHkRRlCgNRAlAvpPCsEx%2BmcCwJBMCRDECWI7mqDVSU8QI7hkQGDig5G6Pubcej6JFUNA8k0bBnoGN0gxYdJhoMZGXjgjGSYjFJVwbXiUL0cJmodxOewquKSZnPRmmyaGBndP3dT6aKQJuZo5j%2B1WDKDQ9D5b2NYaXiaOWpmElCd3JO9eMglnVQKlimL2DVDm1FSRcC2IBga%2BSmvk1z9kXAwKaJEDTFaWmTNaH21IdzYbgxgCjSpkUcIQ7dGZsC2XeKqD4Q9%2BcspI1OjV9kzdkp-Yxl18UOUw1DbbDpK6b9vKzXHT9z3g2XCdr54G7XXHcibzszejzYtYH5S4lR-2nRdN306w6qo71Wum%2B0GwMeFmO66rwP7dn42tCcUfpmX4Osfgvf9EWkM5HwRBBFW9aYE27a9vzABhUhRGzKAwFessrozW6KHu3Mj1CwvVLO9EQn1qw-TrP9BsdMG59xIlnYIJ8GjZVQkydWMw9LWzpCXE4nhbCkkCtoHyZoAQ6hOEJLBORNYYIdngryC447mkdpTYCKsmhmBiIzDWNDtYDklK8XChd8a8TLnwjuCIF4IyYk8Q4ugjDzicIQ1hCD8gDBwZMBhtgbAqObIGcOqCJTYOMP2cK1DpiejJCIlC7D0YPkhH3fQScTLMMkqMUxFiNGqxOATOkE09EzXaOjIOXgEHvg0gaFB%2B8YiwU-PpKE%2B4pylCcabUkS9si-jZsaNuXJwZoK6E47u9gUFgwcbyIpDlQI%2BgUbwnmOCDEzVqV42huDd7tP8fDdGFodRmKYu2UwXSqJ9KAgMr8wx5qw1CbPAYaV%2B7U2DGGJZcgoCvyoHmUAa0NpbV2vQAA1AIHaazP6gJurga6mZ-45g2QWZ6xYv5gMrF9Gsv16yqDgdLbiaNLQUJCZLGRKUHABBLgqAYzsxFeGFhNEeqxwUV06X84Wxdvm-K0epKx1QKYGGBQyMFHMOEN2FrYYmYceYRXtJMYliLVLrx%2Bbg9FALs5Aq5ohPFhVmK2iJUiPeZKIWUt3tMpeKKxTAoZZjQF2KWULDZZIgl9giUYR5XCjldIpoH10ig9VLxtwozRnAqUFjkGivtuysEqURZnzkHsyM8YTlvTOSAdMDq7rXKAXcks9qBDgKrN9Wsf0AZ7FfD8duml176xlsFPUY46HaDOPOR4zdBgAV9nOTxhUG6xpKEQrV8C3xYVDdEqZmMJqlJkFYS8zdXRx22BY6InQgb1C1dYktdxLUgFjPgUggh%2BD4DAGgGgNynWXOzA9fMT0iwerLF6p5kC-VvNKOuKN5qfxPAaDpZODgk5%2Bi3O4dJ0JMadipZcWwPgiFnHlTCTd%2B7k27qvTpdq%2B7vYcwIUKQJa6L0cy3cYHdabnH7oCWQ2S9wuG9X6v4vwSc7YTQXBBq4yiz3-svQYL9N7f1fqpZ2J9pqrgmQuFB2DdQhZnAuAR%2BDgV11Ib-TpVDlT73KMA8%2BnDJRcMlEsqqsaY9CMBB0XFTjZHmwUc-den9tH%2BbcYY9hkDR6CFscpdxnUkHVW6PmEnfjNxBPspQyJvd96uoSckRcQw0nWNgcpSxxepnuMJVU1m8jH7NOPq1poHB66TCYesZJljXHQMWe8wElTcHbMCfs5Ir9tEtAuf5kZ-TBqQMLnw5xu2GGaQ2bfeexz7QwtOciyUejOMgOGeYyUfBlneO%2BaS-JvjQX1Mhcy45m2DkBmufEvlxjcW3OsgRTBxLqquoWlSwhjToX6vivkk1qLLWMVtaM0Zkr3mBbldVZV3zan30ZbvRsBrdlxu5cm4y2LM3QNTT88RnzinKU%2BAC2YVb6WN3IcfcY%2BZBphllv7KOb425L2Zr836AYqxMS1CQ03dNWl9BnDtoEDUQ0eg2CsDSs2%2BJY5OlByYKwewpycLtpmxHdlUhz2rfYKwFpwd0gkMrb9hqsHedBiKZIwPcrMRMnbCQ26TF-HCEhMpNimjbixcBPDYpUc9CpP85wCjvHjNK5mnii74QF0mIjvGrwSeWlRxWtBtlOYVNGDYSYVhRxHkGaVk0SuTR3ZVUBbjg3asbeZU9zh7Q6no720btHWFsl64GBjwY327YRu96JSJYwnGrcaUOM1KFpEilXaMbTV6UjdjZ37ajvCq3NEw2hXnTGht1eo3H%2B7dHAtpbTXPCP9oM8-u0vndbBfNt2mcwaZr13qtGJ18GnEUJOxmRKPzAk4ti%2BSvt57do1pxZAXSD3eule1hxZt7Xu3UaXM3d0mbEPrSU-ULT%2BXqfhcFTV-N1lrbY3zbyrOKHyZjbohjLbr9-xexVh8-V2OD9ajE0qw3U4vnAz3OfOiDZldgUC6Y4%2BQy6NSrCOetuNGOmp%2BLuWGBm2eAuQsBGUGp2K2LeHiEB8%2BUBtuGGMW8EcWiBJWi2RGZWA24Bc%2BVGrOwe0Bu27mBWCBpWlwxBJwbmCmReCGt2lGWmK%2BNBU0dB02oGfmyBqqh2aBxenBQmeePBOBemrWnmghCWzBxWZ27B5BNelBactOjeMB-B8h8W4GPWlKScpG6Ba2B%2BI2vwDeDuRWuh8BBBjBwhlK1mqhzYEhDmFsWkVhw%2BtBeB2U9hQhhh3GPg-WLhQSAYf4Ze88biMeUkqeZeM%2Bl6n%2BjWqoP%2Bu%2BV6A%2BF%2BrhuuYBrhJeOcUaLUmEWes%2B6hh%2Bo28IO2D6zeGRwSWRLWcklGdedk1S0eqicRtasBHmdhh2J6PejBp2TB523GveVWNRc07QbaogNAKANyWyd8Oyj8oAqgNA%2BAogsYQYN82yD8%2ByNqCY0xUA3AH8yYP8zqVyo6oA46ICnqH0PqLy0Caglh9%2BHRbko%2BhUwOSwXSg%2B9SJe5iT2ViEeecTGABg%2BnMSRiwfEWevs-YiQs0Qs0R9QdKlivyr%2B-Cde%2B47M2UWiccmacBcyka-MFwScrQykQ%2BmI1BDk4iwEXECocy7xgiKJbcHUeR4sK8vK8KZuuuXSZgPSQe7MUJkKJJroCab6SJdIKJA8aJD%2BvEWc3uEUWJYpAwbeeUdCIJ5qYJ94waFwZGViIqzJqJPiek5Kaw8pzMipbCkpwagexEsp5sGoJJi%2B2e2paMopViI%2Bd%2BpgeOyq7%2BzCTK7OGJopLC%2B4deCchpLux2BCvygYb6dKOpiJvpB4-pLRLCLpkZbJFuLRP4-MRCiJdOtgba1qcYexpyJxw6ACNylx9ypy06ECvqryMC7ygaJeESxERqCJqK6uXh5Oi6KpdiLx8csybC1KYSeaeUBaeq5MXkKCMZORfsIZYwXsaZ7%2BlQXp3I0sYO9ewSTaCoYM45kSe8v4nyicsUEiRpNJkScBGOpEs0OawKmgZ8SyD558k6VAfaFAeAsYSYg6Fyf8I6gCY6wClZ1x3qzyUC-qDYoy55BRBpgJ24oa8KViUp0c65qu6MNkE8mhBgTkNKMcsFGFCRhcA5RomaIpoSioVwvyMkjcF4pUM5jukFBK84oEYMaFDkk8SkkKLk08Ruzei4%2BSDpIJ7QqQsJJwJwAc4pjZfZ4lduAIPJDCyZbEf6ypsB7ia6RJCp-ZkFWk%2BF24hFPuXZ4cE%2BPpGwf29F3Iw8qwk4Y%2B0E-E5c7c4i4cYyQ8kKfusqNF4qB4lQFlLWUKnFSlC5KciaikXsjJmEK8tcSa0Jm8ocuF4W2lMqlQzCj2UFppOMQ0dewVdUPldIROzKP4IBekfujOtlIu9u65RMBpeU3omsoBToJl1pZlVSdEyVPo%2BVNCQZEs-wLoSZDpKV7VncfYTE9JnxRVDRie2sZVI1nsLlGw4aRcrMTV6iFVhsblMFdVg8Z5uRWVCufe7ojKeJXlpQWV92fl9p6FtEwVXEhJBsA1HJbCnycFFuGadw%2BlhVdSFMrpOSL14sIZeUU4KVW8ekdCq1mlEKZuGodUwNLZEKvc0Va1Wle%2BgW%2BimRs0JF9eTx-ciMOMeExiOlgVYpTu0lgl1WbanAiAL5WY75n5JZP5ZZbqE6Dy5YwFs6dZDxAISszxAirRfVmu%2BRoOGcdEwlC%2BlJXS0uLFjUGFjkHF8e7c8axVmhxFRCEF9V3ZAtPJq88kgi41u8Xun1siAtVVXxwtgucttECt4RaNplKtVi2Nrx4IJU1tDkmt4c2tFiPQmtEIIG8lqCwUQp2IWEjZvcVpPNeNdk-NcWhIq67Eam95QAA) and it always works, I don't really know what you might be doing differently, whenever `params.slug` changes it re-runs the `load` function and updates the `data` value in `+page.svelte` without needing to declare a reactive variable.

PS: Whenever possible, use `fetch` from `load( {fetch, params} )` instead of regular `fetch`

**Reply by eallenOP** [reply]
I worked it out. In your repro you render the page straight from `data` here
```svelte
<script>
   export let data
</script>
<h1>
   {data.user.username}
</h1>
<p>{data.user.phone}</p>
```
but in the OP and in my own code we've grabbed the data out of `data` to make life easier in the rest of the `+page.svelte` file, like this:
```svelte
<script>
   export let data
   let user = data.user;
</script>
<h1>
   {user.username}
</h1>
<p>{user.phone}</p>
```
but because `user` isn't reactive it wasn't updating with the new user even when `data` changed, hence looking as though the page doesn't rerender. So I'm not sure whether it would be affected at all by not using `params.slug` in the load function after all.

Hence the solution accepted above is still the solution:
```svelte
<script>
   export let data
   $: user = data.user;
</script>
<h1>
   {user.username}
</h1>
<p>{user.phone}</p>
```

**Reply by ticup** [reply]
Thanks for the explanation @eallenOP , I have been struggling with this for a while, sometimes it was changing, sometimes it wasn't. Now I know why. I feel like we should have better documentation on such subtle things.

**Comment by Anoopvempeny** [comment]
gft

**Comment by amirzezo201** [comment]
did anyone try to make it work in Runes mode

**Reply by fo-rk** [reply]
I don't know if it's wrong for other reasons, but I used:
```js
let { data } = $props();
let user = $derived(data.user);
```

**Reply by data-miner00** [reply]
It's so weird that this still happens in Runes mode.

**Reply by data-miner00** [reply]
For my case, found out that SvelteKit caches the array between [slug] pages while navigating and `onMount` is only called once during the first page.

To circumvent this issue, I used a combination of `$effect` and `page.params.slug` (where slug is the self-defined parameter) to update the cached array by calling the update method.

```ts
import { page } from '$app/state';

let array = $state<string[]>([]);

$effect(() => {
	if (page.url.slug) {
		fetchAndUpdateData(page.params.slug);
	}
});

async function fetchAndUpdateData(slug: string) {
	array = await fetchFromDb(slug) as string[];
}
```
